src/API/Google/AdsCampaign.php

```diff
@@ -18,54 +20,63 @@
 use Google\Ads\GoogleAds\V9\Services\CampaignOperation;
 use Google\Ads\GoogleAds\V9\Services\CampaignServiceClient;
 use Google\Ads\GoogleAds\V9\Services\GoogleAdsRow;
-use Google\Ads\GoogleAds\V9\Services\MutateCampaignResult;
+use Google\Ads\GoogleAds\V9\Services\MutateOperation;
 use Google\ApiCore\ApiException;
 use Google\ApiCore\ValidationException;
 use Exception;
 
 /**
- * Class AdsCampaign
+ * Class AdsCampaign (Smart Shopping Campaign)
+ * https://developers.google.com/google-ads/api/docs/smart-campaigns/overview
+ *
+ * ContainerAware used for:
+ * - AdsGroup
+ *
+ * @since x.x.x Refactored to use batch requests when operating on campaigns.
  *
  * @package Automattic\WooCommerce\GoogleListingsAndAds\API\Google
  */
-class AdsCampaign implements OptionsAwareInterface {
+class AdsCampaign implements ContainerAwareInterface, OptionsAwareInterface {
 
 	use ApiExceptionTrait;
+	use ContainerAwareTrait;
 	use OptionsAwareTrait;
 	use MicroTrait;
 
 	/**
-	 * The Google Ads Client.
+	 * Temporary ID to use within a batch job.
+	 * A negative number which is unique for all the created resources.
 	 *
-	 * @var GoogleAdsClient
+	 * @var int
 	 */
-	protected $client;
+	protected const TEMPORARY_ID = -1;
 
 	/**
-	 * @var AdsCampaignBudget $ads_campaign_budget
+	 * The Google Ads Client.
+	 *
+	 * @var GoogleAdsClient
 	 */
-	protected $ads_campaign_budget;
+	protected $client;
 
 	/**
-	 * @var AdsGroup $ads_group
+	 * @var AdsCampaignBudget $budget
 	 */
-	protected $ads_group;
-
+	protected $budget;
 
 	/**
 	 * AdsCampaign constructor.
 	 *
 	 * @param GoogleAdsClient   $client
-	 * @param AdsCampaignBudget $ads_campaign_budget
-	 * @param AdsGroup          $ads_group
+	 * @param AdsCampaignBudget $budget
 	 */
-	public function __construct( GoogleAdsClient $client, AdsCampaignBudget $ads_campaign_budget, AdsGroup $ads_group ) {
-		$this->client              = $client;
-		$this->ads_campaign_budget = $ads_campaign_budget;
-		$this->ads_group           = $ads_group;
+	public function __construct( GoogleAdsClient $client, AdsCampaignBudget $budget ) {
+		$this->client = $client;
+		$this->budget = $budget;
```

Comment:
❔ Curious that since we were using `ads_group` from the container, could the `budget` be extracted from the container as well so we wouldn't need to pass the ads campaign budget when initialising AdsCampaign?

Reply:
It could. But the reason I added it as a direct dependency is it's used by most of the functions (direct dependencies are preferred). The Ad Group could have also been a direct dependency, but in preparation for the PMax changes, we no longer use an Ad Group but an Asset Group instead, so I retrieve those from the Container only when they are needed, depending on Campaign type.